src/api_rework.jl

```diff
@@ -0,0 +1,120 @@
+# I am putting everything in here for now. The contents of this file should be moved around in the future.
+"""
+    findpeaks(x) -> NamedTuple
+
+Find the peaks in a vector `x`. 
+A `NamedTuple` is returned with the original vector 
+in the field `data`, and the indices of the peaks 
+in the field `inds`.
+
+This function serves as the entry-point for other 
+functions such as `peakproms!` and `peakwidths!`
+"""
+function findpeaks(x::AbstractVector)
+    data = x
+    inds, _ = findmaxima(x)
+    return (data=x, inds=inds)
+end
+
+"""
+    _filter_fields!(pks, new_inds)
+
+Internal function to mutate the vectors in all fields of `pks` 
+to remove any element with a corresponding `pks.inds` entry 
+not present in `new_inds`.
+"""
+function _filter_fields!(pks, new_inds)
+    # Add checks on `pks` to see if it has fields "data" and "inds"?
+    mask = pks.inds .∉ Ref(new_inds)
+    for field in filter(!=(:data), propertynames(pks)) # Do not want to mutate data vector
+        v_to_be_mutated = getfield(pks, field)
+        if length(v_to_be_mutated) != length(mask)
+            error("Internal error: The length of the vector in field `$field` is $(length(v_to_be_mutated)), but was expected to be $(length(mask))")
+        end
+        deleteat!(getfield(pks, field), mask)
+    end
+    return nothing
+end
+
+"""
+    peakproms!(pks::NamedTuple; min=0, max=Inf, strict=true)
+    peakproms!(; min=0, max=Inf, strict=true)
+
+Find the prominences of the peaks in `pks`, and filter out any peak 
+with a prominence smaller than `min` or greater than `max`.
+The prominences are returned in the field `:proms` of the returned named tuple.
+
+If the positional argument `pks` is omitted, an anonymus function is returned 
+that performs the action (adding field `:proms` and filtering) to its input.
+"""
+function peakproms!(pks::NamedTuple; min=0, max=Inf, strict=true)
+    if !hasproperty(pks, :proms)
+        # Avoid filtering by min/max/strict here, so that it always happens outside if-statement.
+        # Pro: one less edge case. Con: More internal allocations
+        _, proms = _peakproms(pks.inds, pks.data)
+        pks = merge(pks, (; proms))
+    end
+    new_inds, _ = _peakproms(pks.inds, pks.data; minprom=min, maxprom=max, strict)
+    _filter_fields!(pks, new_inds)
```

Comment:
I'm not sure I fully understand your logic flow here. Would something like this work instead? (and/or explain why and what's going on if not)

```suggestion
    if !hasproperty(pks, :proms)
        # Avoid filtering by min/max/strict here, so that it always happens outside if-statement.
        # Pro: one less edge case. Con: More internal allocations
        _, proms = _peakproms(pks.inds, pks.data; strict)
        pks = merge(pks, (; proms))
    end

    if !isnothing(min) || !isnothing(max)
        lo = something(min, zero(eltype(x)))
        up = something(max, typemax(Base.nonmissingtype(eltype(x))))
        matched = findall(x -> !ismissing(x) && !(lo ≤ x ≤ up), pks.proms)
        deleteat!(pks.peaks, matched)
        deleteat!(pks.proms, matched)
    end
```
It looks like you're only using the second `peakproms` call to get the results of the filtering by min/maxprom, so it would be much (computationally) cheaper to just copy/paste that section of the function than call the whole thing again.

Reply:
Right, this is the type of internal cleanup I mentioned in the original post. I am aware I sinned significantly in the original draft xD

The reason I avoided the `strict` keyword in the first call is because I believed that it would `filter` peaks. I can see in the readme example that the only effect regarding the existance of a peak is that with `strict=false`, a peak is added in one case. The important thing is that peaks should not be added/removed in the initial call - the only effect should be adding the new field to the named tuple. But I have not concidered that peaks may be added. Would it make sense to have a field `pks.strict` that would ensure that all functions would treat NaN/missing with the same strictness, in order to avoid adding peaks at some point? I think that would break things silently, unless we handle it cleverly.

Then, the filtering happens outside the loop, as you nicely suggested. This is the step where we are allowed to remove peaks, with the key point being that we do the same deletion for all fields of the named tuple (excluding `data`). In your suggestion, we only remove peaks for the index-field and the just-computed field. It is important that if the named tuple had a field `heights`, it would say in sync. As this step is the same for all functions, my original implementation using `_filter_fields!` makes more sense. So what about something like this:
```julia
function peakheights!(pks::NamedTuple; min=0, max=Inf)
    if !hasproperty(pks, :heights)
        # Avoid filtering by min/max here, so that it always happens outside if-statement.
        # Pro: one less edge case. Con: More internal allocations
        heights = pks.data[pks.inds]
        pks = merge(pks, (; heights))
    end
    if !isnothing(min) || !isnothing(max)
        lo = something(min, zero(eltype(x)))
        up = something(max, typemax(Base.nonmissingtype(eltype(x))))
        matched = findall(x -> !ismissing(x) && !(lo ≤ x ≤ up), pks.proms)
        _filter_fields!(pks, matched)
    end
    return pks
end
```

Having rewritten `_filter_peaks!` to something like
```julia
function _filter_fields!(pks, mask)
    # Add checks on `pks` to see if it has fields "data" and "inds"?
    for field in (:inds, :proms, :heights, :widths, :edges)  # Only risk mutating fields added by this package
        hasfield(pks, field) || continue  # Do nothing if field is not present
        v_to_be_mutated = getfield(pks, field)
        if length(v_to_be_mutated) != length(mask)  # Some performance overhead, but good test to have. Remove once tests are in place?
            error("Internal error: The length of the vector in field `$field` is $(length(v_to_be_mutated)), but was expected to be $(length(mask))")
        end
        deleteat!(v_to_be_mutated, mask)
    end
    return nothing
end
```

<details>

<summary>Sidenote about performance, showing that deleting all entries in the mask in a loop is fastest:</summary>

```
julia> using BenchmarkTools

julia> function f(N, length, criterion = >(0.5))
               # Test with simultaneous deletion
               pks1 = ([rand(length) for _ in 1:N]..., )
               pks2 = deepcopy(pks1)
               @time "Deleting rows     " for i in reverse(eachindex(pks1[1])) # Reverse to not change index of yet-to-filter features
                       if criterion(pks1[1][i])
                               deleteat!.(pks1, Ref(i))
                       end
               end
               
               @time "Deleting columns 1" begin
                       mask = criterion.(pks2[1])
                       deleteat!.(pks2, Ref(mask))
               end
       
               @time "Deleting columns 2" begin
                               mask = criterion.(pks2[1])
                               for i in eachindex(pks2)
                                       deleteat!(pks2[i], mask)
                               end
                       end
       
               @show pks1 .== pks2
               
               return pks1, pks2
       end
f (generic function with 2 methods)

julia> f(10, 100);
Deleting rows     : 0.000085 seconds (504 allocations: 38.500 KiB)
Deleting columns 1: 0.000013 seconds (11 allocations: 816 bytes)
Deleting columns 2: 0.000002 seconds (2 allocations: 96 bytes)
pks1 .== pks2 = (true, true, true, true, true, true, true, true, true, true)

julia> f(10, 1000);
Deleting rows     : 0.003871 seconds (4.64 k allocations: 354.750 KiB)
Deleting columns 1: 0.000114 seconds (12 allocations: 5.094 KiB)
Deleting columns 2: 0.000030 seconds (3 allocations: 4.344 KiB)
pks1 .== pks2 = (true, true, true, true, true, true, true, true, true, true)

julia> f(10, 10000);
Deleting rows     : 0.047478 seconds (45.23 k allocations: 3.374 MiB)
Deleting columns 1: 0.000165 seconds (12 allocations: 6.234 KiB)
Deleting columns 2: 0.000061 seconds (3 allocations: 4.938 KiB)
pks1 .== pks2 = (true, true, true, true, true, true, true, true, true, true)
```

</details>

Reply:
`strict==false` will only add peaks in `findmaxima/minima` functions. For `peakproms` and `peakwidths`, `strict==false` will change whether a prominence or width is calculable. tl;dr: `strict` doesn't change filtering, so `strict` doesn't need to be included in the NamedTuple.

Otherwise, yeah what you have there would work.

Reply:
I now filter directly, as opposed to calling the original functions twice. I also make use of the strict keyword in all functions.